File: oop/04-abstraction/typescript/chequeAccount.ts
/*
 * Demonstrates: Abstraction - Concrete Implementation
 * Concept: ChequeAccount implements all abstract methods from Account
 */

import { Account } from "./account";

export class ChequeAccount extends Account {
    public readonly overdraftLimit: number;
    private _transactionCount: number = 0;
    private static readonly FEE_PER_TRANSACTION = 2.50;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number, overdraftLimit: number) {
        super(accountNumber, accountHolder, openingBalance);
        this.overdraftLimit = overdraftLimit;
    }

    get transactionCount(): number {
        return this._transactionCount;
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Withdrawal amount must be greater than zero.");
        }

        const availableFunds = this._balance + this.overdraftLimit;

        if (amount > availableFunds) {
            throw new Error(`Exceeds overdraft limit. Available: R${availableFunds.toFixed(2)}`);
        }

        this._balance -= amount;
        this._transactionCount++;
    }

    calculateMonthlyFees(): number {
        return this._transactionCount * ChequeAccount.FEE_PER_TRANSACTION;
    }

    getAccountSummary(): string {
        return `Cheque Account ${this.accountNumber} | Holder: ${this.accountHolder} | Balance: R${this._balance.toFixed(2)} | Overdraft: R${this.overdraftLimit.toFixed(2)} | Transactions: ${this._transactionCount}`;
    }

    resetTransactionCount(): void {
        this._transactionCount = 0;
    }
}

File: oop/04-abstraction/csharp/ChequeAccount.cs
/*
 * Demonstrates: Abstraction - Concrete Implementation
 * Concept: ChequeAccount implements all abstract methods from Account
 */

namespace FinanceDomain.Abstraction;

public class ChequeAccount : Account
{
    public decimal OverdraftLimit { get; }
    public int TransactionCount { get; private set; }
    private const decimal FeePerTransaction = 2.50m;

    public ChequeAccount(string accountNumber, string accountHolder, decimal openingBalance, decimal overdraftLimit)
        : base(accountNumber, accountHolder, openingBalance)
    {
        OverdraftLimit = overdraftLimit;
        TransactionCount = 0;
    }

    public override void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Withdrawal amount must be greater than zero.");

        decimal availableFunds = _balance + OverdraftLimit;

        if (amount > availableFunds)
            throw new InvalidOperationException($"Exceeds overdraft limit. Available: R{availableFunds:F2}");

        _balance -= amount;
        TransactionCount++;
    }

    public override decimal CalculateMonthlyFees()
    {
        return TransactionCount * FeePerTransaction;
    }

    public override string GetAccountSummary()
    {
        return $"Cheque Account {AccountNumber} | Holder: {AccountHolder} | Balance: R{_balance:F2} | Overdraft: R{OverdraftLimit:F2} | Transactions: {TransactionCount}";
    }

    public void ResetTransactionCount()
    {
        TransactionCount = 0;
    }
}
